File: backend/entity/schedule.go
package entity

import "time"

// AutoScanSchedule เก็บตารางเวลา auto scan ที่ผู้ใช้ตั้งไว้
type AutoScanSchedule struct {
	ID         uint       `gorm:"primaryKey;autoIncrement"`
	TaskID     string     `gorm:"column:task_id;not null"`
	TaskName   string     `gorm:"column:task_name"`
	Frequency  string     `gorm:"column:frequency;not null"` // once | monthly | yearly
	ScanTime   string     `gorm:"column:scan_time;not null"` // "HH:mm" 24-hr
	ScheduleAt *time.Time `gorm:"column:schedule_at"`        // for "once" — full datetime
	DayOfMonth *int       `gorm:"column:day_of_month"`       // for "monthly" (1-31)
	Month      *int       `gorm:"column:month"`              // for "yearly" (1-12)
	Day        *int       `gorm:"column:day"`                // for "yearly" (1-31)
	Enabled    bool       `gorm:"column:enabled;default:true"`
	LastRunAt  *time.Time `gorm:"column:last_run_at"`
	NextRunAt  *time.Time `gorm:"column:next_run_at"`
	CreatedAt  time.Time
	UpdatedAt  time.Time
}

func (AutoScanSchedule) TableName() string {
	return "auto_scan_schedules"
}

File: frontend/src/services/schedule.ts
import axios from "axios";
import { apiUrl, installMaintenanceInterceptor } from "./api";

const scheduleApi = axios.create({
  baseURL: apiUrl,
  withCredentials: true,
  timeout: 15000,
  headers: { "Content-Type": "application/json", "ngrok-skip-browser-warning": "true" },
});

installMaintenanceInterceptor(scheduleApi);

// ── Types ──────────────────────────────────────────────────────────────────

export type ScheduleFrequency = "once" | "monthly" | "yearly";

export type AutoScanScheduleDTO = {
  id: number;
  task_id: string;
  task_name: string;
  frequency: ScheduleFrequency;
  scan_time: string;          // "HH:mm"
  timezone: string;           // IANA timezone, e.g. "Asia/Bangkok"
  schedule_at?: string;       // "YYYY-MM-DD" — for once
  day_of_month?: number;      // 1-31 — for monthly
  month?: number;             // 1-12 — for yearly
  day?: number;               // 1-31 — for yearly
  enabled: boolean;
  last_run_at?: string;       // ISO
  next_run_at?: string;       // ISO
  created_at: string;
};

export type CreateScheduleRequest = {
  task_id: string;
  task_name: string;
  frequency: ScheduleFrequency;
  scan_time: string;
  timezone: string;           // IANA timezone, e.g. "Asia/Bangkok"
  schedule_at?: string;
  day_of_month?: number;
  month?: number;
  day?: number;
};

// ── API calls ──────────────────────────────────────────────────────────────

export const ListScanSchedules = async (): Promise<AutoScanScheduleDTO[]> => {
  try {
    const res = await scheduleApi.get("/scan-schedules");
    const data = res.data?.data ?? res.data;
    return Array.isArray(data) ? (data as AutoScanScheduleDTO[]) : [];
  } catch (err) {
    console.error("ListScanSchedules error:", err);
    return [];
  }
};

export const CreateScanSchedule = async (
  req: CreateScheduleRequest,
): Promise<AutoScanScheduleDTO> => {
  const res = await scheduleApi.post("/scan-schedules", req);
  return res.data as AutoScanScheduleDTO;
};

export const UpdateScanSchedule = async (
  id: number,
  enabled: boolean,
): Promise<AutoScanScheduleDTO> => {
  const res = await scheduleApi.patch(`/scan-schedules/${id}`, { enabled });
  return res.data as AutoScanScheduleDTO;
};

export const DeleteScanSchedule = async (id: number): Promise<void> => {
  await scheduleApi.delete(`/scan-schedules/${id}`);
};
